In [1]:
import numpy as np
import pandas as pd
import re

In [2]:
pd.options.display.float_format = '{:.2f}'.format

### Companies Data Analysis

In [3]:
df_com = pd.read_csv('companies.txt', sep='\t', encoding='ISO-8859-1')

In [4]:
df_com.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB


In [5]:
df_com.describe()

permalink   name                homepage_url  \
count                          66368  66367                       61310   
unique                         66368  66102                       61191   
top     /Organization/Mission-Motors  Blink  http://www.askforoffer.com   
freq                               1      4                           5   

       category_list     status country_code state_code       region  \
count          63220      66368        59410      57821        58338   
unique         27296          4          137        311         1092   
top         Software  operating          USA         CA  SF Bay Area   
freq            3995      53034        37601      12900         8804   

                 city  founded_at  
count           58340       51147  
unique           5111        3978  
top     San Francisco  01-01-2012  
freq             3526        2730

In [6]:
df_com.size

663680

In [7]:
df_com.shape

(66368, 10)

In [8]:
df_com.columns

Index(['permalink', 'name', 'homepage_url', 'category_list', 'status',
       'country_code', 'state_code', 'region', 'city', 'founded_at'],
      dtype='object')

In [9]:
df_com.head(10)

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   
5     /Organization/01Games-Technology      01Games Technology   
6  /Organization/0Ndine-Biomedical-Inc  Ondine Biomedical Inc.   
7                 /Organization/0Xdata                  H2O.ai   
8                      /Organization/1                One Inc.   
9            /Organization/1-2-3-Listo             1,2,3 Listo   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   
5             http://www.01games.hk/   
6               http://ondinebio.com   
7                     http://h2o.ai/   
8                 http://whatis1.com   
9            http://www.123listo.com   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   
5                                              Games  operating          HKG   
6                                      Biotechnology  operating          CAN   
7                                          Analytics  operating          USA   
8                                             Mobile  operating          USA   
9                                         E-Commerce  operating          CHL   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010  
5        NaN              Hong Kong      Hong Kong         NaN  
6         BC              Vancouver      Vancouver  01-01-1997  
7         CA            SF Bay Area  Mountain View  01-01-2011  
8         CA            SF Bay Area  San Francisco  01-08-2011  
9         12               Santiago     Las Condes  01-01-2012

If we observe the permalink, the case is not uniform; lets convert into lower case

In [10]:
df_com['permalink'] = df_com['permalink'].str.lower()

In [11]:
df_com.head(10)

permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
3                /organization/0-6-com                 0-6.com   
4       /organization/004-technologies        004 Technologies   
5     /organization/01games-technology      01Games Technology   
6  /organization/0ndine-biomedical-inc  Ondine Biomedical Inc.   
7                 /organization/0xdata                  H2O.ai   
8                      /organization/1                One Inc.   
9            /organization/1-2-3-listo             1,2,3 Listo   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   
5             http://www.01games.hk/   
6               http://ondinebio.com   
7                     http://h2o.ai/   
8                 http://whatis1.com   
9            http://www.123listo.com   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   
5                                              Games  operating          HKG   
6                                      Biotechnology  operating          CAN   
7                                          Analytics  operating          USA   
8                                             Mobile  operating          USA   
9                                         E-Commerce  operating          CHL   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010  
5        NaN              Hong Kong      Hong Kong         NaN  
6         BC              Vancouver      Vancouver  01-01-1997  
7         CA            SF Bay Area  Mountain View  01-01-2011  
8         CA            SF Bay Area  San Francisco  01-08-2011  
9         12               Santiago     Las Condes  01-01-2012

## As per the description of the Companies data, Permalink is unique to each company

## Answer to the Question - How many unique companies are present in the companies file?

In [12]:
unique_permalink = df_com.permalink.unique()
print(len(unique_permalink))

66368


In [13]:
df_com['name'] = df_com['name'].str.lower()

In [14]:
df_com.head(15)

permalink                         name  \
0                        /organization/-fame                        #fame   
1                     /organization/-qounter                     :qounter   
2        /organization/-the-one-of-them-inc-       (the) one of them,inc.   
3                      /organization/0-6-com                      0-6.com   
4             /organization/004-technologies             004 technologies   
5           /organization/01games-technology           01games technology   
6        /organization/0ndine-biomedical-inc       ondine biomedical inc.   
7                       /organization/0xdata                       h2o.ai   
8                            /organization/1                     one inc.   
9                  /organization/1-2-3-listo                  1,2,3 listo   
10                     /organization/1-4-all                      1-4 all   
11            /organization/1-618-technology             1.618 technology   
12               /organization/1-800-dentist                1-800-dentist   
13               /organization/1-800-doctors                1-800-doctors   
14  /organization/1-800-publicrelations-inc-  1-800-publicrelations, inc.   

                          homepage_url  \
0                   http://livfame.com   
1               http://www.qounter.com   
2                  http://oneofthem.jp   
3                   http://www.0-6.com   
4    http://004gmbh.de/en/004-interact   
5               http://www.01games.hk/   
6                 http://ondinebio.com   
7                       http://h2o.ai/   
8                   http://whatis1.com   
9              http://www.123listo.com   
10                                 NaN   
11  http://www.Homeandcondogallery.com   
12          http://www.1800dentist.com   
13              http://1800doctors.com   
14  http://www.1800publicrelations.com   

                                        category_list     status country_code  \
0                                               Media  operating          IND   
1   Application Platforms|Real Time|Social Network...  operating          USA   
2                                   Apps|Games|Mobile  operating          NaN   
3                                         Curated Web  operating          CHN   
4                                            Software  operating          USA   
5                                               Games  operating          HKG   
6                                       Biotechnology  operating          CAN   
7                                           Analytics  operating          USA   
8                                              Mobile  operating          USA   
9                                          E-Commerce  operating          CHL   
10                       Entertainment|Games|Software  operating          USA   
11                 Networking|Real Estate|Web Hosting  operating          USA   
12                                Health and Wellness  operating          USA   
13                                Health and Wellness  operating          USA   
14          Internet Marketing|Media|Public Relations  operating          USA   

   state_code                 region               city  founded_at  
0          16                 Mumbai             Mumbai         NaN  
1          DE             DE - Other      Delaware City  04-09-2014  
2         NaN                    NaN                NaN         NaN  
3          22                Beijing            Beijing  01-01-2007  
4          IL  Springfield, Illinois          Champaign  01-01-2010  
5         NaN              Hong Kong          Hong Kong         NaN  
6          BC              Vancouver          Vancouver  01-01-1997  
7          CA            SF Bay Area      Mountain View  01-01-2011  
8          CA            SF Bay Area      San Francisco  01-08-2011  
9          12               Santiago         Las Condes  01-01-2012  
10         NC             NC - Other  Connellys Springs       

In [15]:
unique_name = df_com.name.unique()
len(unique_name)

66038

In [16]:
df_dup_com = df_com[df_com.duplicated(['name'])]

In [17]:
df_dup_com.head()

permalink    name               homepage_url  \
282    /organization/3divaz-3  3divaz  http://www.3divaz.ch/Home   
1527   /organization/adtena-2  adtena          http://adtena.com   
1996    /organization/agora-6   agora       https://www.agora.co   
2460    /organization/alike-2   alike           http://alike.fr/   
3006  /organization/amicus-co  amicus       http://www.amicus.co   

                     category_list     status country_code state_code  \
282                            NaN  operating          CHE          1   
1527                           NaN     closed          NaN        NaN   
1996  Android|Apps|Internet|Mobile  operating          USA         NY   
2460                        Design     closed          FRA         A8   
3006   Apps|Mobile|Mobile Commerce  operating          IND          7   

             region       city  founded_at  
282     CHE - Other    Wildegg  01-07-2014  
1527            NaN        NaN  15-11-2014  
1996  New York City   New York  01-07-2013  
2460          Paris      Paris         NaN  
3006      New Delhi  New Delhi  01-01-2015

In [18]:
# df_dup_com1 = df_com[df_com.duplicated()]

In [19]:
# df_dup_com1.shape

In [20]:
df_dup_com.shape

(330, 10)

There are around 330 companies which are duplicated; lets filter that later and lets continue using the company table
with duplicates

In [21]:
df_com.iloc[281]

permalink           /organization/3divaz-2
name                                3divaz
homepage_url     http://www.3divaz.ch/Home
category_list                          NaN
status                              closed
country_code                           NaN
state_code                             NaN
region                                 NaN
city                                   NaN
founded_at                      01-07-2014
Name: 281, dtype: object

In [22]:
df_com.iloc[282]

permalink           /organization/3divaz-3
name                                3divaz
homepage_url     http://www.3divaz.ch/Home
category_list                          NaN
status                           operating
country_code                           CHE
state_code                               1
region                         CHE - Other
city                               Wildegg
founded_at                      01-07-2014
Name: 282, dtype: object

In [23]:
# TODO - display all duplicated companies

## Rounds

In [24]:
df_rounds = pd.read_csv('rounds2.csv', encoding='ISO-8859-1')

In [25]:
df_rounds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [26]:
df_rounds.describe()

raised_amount_usd
count           94959.00
mean         10426869.33
std         114821247.98
min                 0.00
25%            322500.00
50%           1680511.00
75%           7000000.00
max       21271935000.00

In [27]:
df_rounds.shape

(114949, 6)

In [28]:
df_rounds.columns

Index(['company_permalink', 'funding_round_permalink', 'funding_round_type',
       'funding_round_code', 'funded_at', 'raised_amount_usd'],
      dtype='object')

In [29]:
df_rounds.head(10)

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   
5       /ORGANIZATION/004-TECHNOLOGIES   
6     /organization/01games-technology   
7  /ORGANIZATION/0NDINE-BIOMEDICAL-INC   
8  /organization/0ndine-biomedical-inc   
9                 /ORGANIZATION/0XDATA   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
5  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
6  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
7  /funding-round/2b9d3ac293d5cdccbecff5c8cb0f327d               seed   
8  /funding-round/954b9499724b946ad8c396a57a5f3b72            venture   
9  /funding-round/383a9bd2c04f7038bb543ccef5ba3eae               seed   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015        10000000.00  
1                  A  14-10-2014                nan  
2                NaN  01-03-2014          700000.00  
3                  B  30-01-2014         3406878.00  
4                  A  19-03-2008         2000000.00  
5                NaN  24-07-2014                nan  
6                NaN  01-07-2014           41250.00  
7                NaN  11-09-2009           43360.00  
8                NaN  21-12-2009          719491.00  
9                NaN  22-05-2013         3000000.00

In [ ]:
df_rounds['company_permalink'] = df_rounds['company_permalink'].str.lower()

In [ ]:
df_rounds.head(10)

In [ ]:
df_rounds.company_permalink.shape

## Answer to the Question - How many unique companies are present in rounds2?

In [ ]:
unique_comp_perm = df_rounds.company_permalink.unique()
print(len(unique_comp_perm))

## Analysis - Q.Are there any companies in the rounds2 file which are not  present in companies ? Answer Y/N.

As per the below calculation, there are 2 companies missing in companies data compared to Rounds data.
Also the Missing comapnies listed

In [ ]:
set_company_permlink = set(unique_comp_perm)

In [ ]:
set_permlink = set(unique_permalink)

In [ ]:
len(set_company_permlink)

In [ ]:
len(set_permlink)

In [ ]:
# 2 companies missing in companies table compared to the rounds table

#### Data present in Rounds table which isnt present in companies table

In [ ]:
set_company_permlink.difference(set_permlink)

#### Another Way

In [ ]:
set_diff_df = pd.concat([df_rounds['company_permalink'], df_com['permalink'], df_com['permalink']]).drop_duplicates(keep=False)
print(set_diff_df)

In [ ]:
# Even though the diff is 2; the actual difference is 7 (len(set_diff_df))
# so after merge 114949-7 rows should be pesent in the merged df

In [ ]:
# data in companies which isnt present in rounds
set_permlink.difference(set_company_permlink)

#### merge the data frames
Need to complete the information of rounds from the company df. So have to join both the df based on the intersection values 

In [ ]:
master_frame = pd.merge(df_rounds, df_com, how='inner', left_on='company_permalink', right_on='permalink')

In [ ]:
master_frame.head(15)

In [ ]:
master_frame.info()

### [Answer] Q . Merge the two data frames so that all  variables (columns)  in the companies frame are added to the rounds2 data frame. Name the merged frame master_frame. How many observations are present in master_frame ?

In [ ]:
master_frame.shape

In [ ]:
print(df_rounds.shape)
print(df_rounds.columns)

In [ ]:
print(df_com.shape)
print(df_com.columns)

In [ ]:
master_frame.to_csv('master.csv')

#### DAta clean-up for further analysis

In [ ]:
master_frame.isnull().all()

In [ ]:
master_frame.isnull().any()

In [ ]:
master_frame.isnull().sum()

In [ ]:
round(100*master_frame.isnull().sum()/len(master_frame.index),2)

In [ ]:
# Not used any where - funding_round_code, founded_at, city, state_Code, region, homepage_url

In [ ]:
master_frame.drop('funding_round_code', axis=1, inplace=True)

In [ ]:
master_frame.drop('founded_at', axis=1, inplace=True)

In [ ]:
master_frame.drop('city', axis=1, inplace=True)

In [ ]:
master_frame.drop('state_code', axis=1, inplace=True)

In [ ]:
master_frame.drop('region', axis=1, inplace=True)

In [ ]:
master_frame.drop('homepage_url', axis=1, inplace=True)

In [ ]:
master_frame['permalink'].equals(master_frame['company_permalink'])

In [ ]:
# one can be removed

In [ ]:
master_frame.drop('company_permalink', axis=1, inplace=True)

In [ ]:
round(100*master_frame.isnull().sum()/len(master_frame.index),2)

In [ ]:
# raised_amt = master_frame.raised_amount_usd
# raised_amt.isnull()

In [ ]:
# raised_amt[1]

In [ ]:
master_frame.dropna(subset=['raised_amount_usd'], inplace=True)

In [ ]:
master_frame.head()

In [ ]:
round(100*master_frame.isnull().sum()/len(master_frame.index),2)

In [ ]:
master_frame.isnull().sum()

In [ ]:
master_frame.shape

In [ ]:
master_frame.to_csv('master.csv')

In [ ]:
index_countries = np.where(master_frame['country_code'].isna())

In [ ]:
print(index_countries, len(index_countries[0]))

In [ ]:
index_cat = np.where(master_frame['category_list'].isna())

In [ ]:
print(index_cat, len(index_cat[0]))

In [ ]:
bool_cat = pd.isnull(master_frame['category_list'])
print(len(index_cat))

In [ ]:
# print(bool_cat)

In [ ]:
# categories can have Others as primary_sector; so lets fill the category data

In [ ]:
master_frame.loc[bool_cat, ['category_list']] = "unknown_cat"

In [ ]:
master_frame.isnull().sum()

In [ ]:
bool_country = pd.isnull(master_frame['country_code'])

In [ ]:
master_frame.loc[bool_country, ['country_code']] = "unknown_country"

In [ ]:
round(100*master_frame.isnull().sum()/len(master_frame.index),2)

#### MAster frame is complete.

In [ ]:
master_frame.info()

In [ ]:
master_frame.describe()

### CheckPoint2

In [ ]:
f_mean = master_frame.groupby('funding_round_type').mean()
print(type(f_mean))
f_mean.sort_values(ascending=True, by=['raised_amount_usd'])

In [ ]:
f_mean.info()

In [ ]:
f_mean.columns

In [ ]:
print(master_frame.funding_round_type.unique())

In [ ]:
f_venture = master_frame['funding_round_type']=='venture'
f_seed = master_frame['funding_round_type']=='seed'
f_pequity = master_frame['funding_round_type']=='private_equity'
f_angel = master_frame['funding_round_type']=='angel'

In [ ]:
df_venture = master_frame[f_venture]
df_seed = master_frame[f_seed]
df_pequity = master_frame[f_pequity]
df_angel = master_frame[f_angel]

### Answers to the Question
Average funding amount of venture type

Average funding amount of angel type

Average funding amount of seed type

Average funding amount of private equity type


In [ ]:
print('Venture - ' + str(round(df_venture.raised_amount_usd.mean(),2)))
print('Seed - ' + str(round(df_seed.raised_amount_usd.mean(),2)))
print('Private Equity - ' + str(round(df_pequity.raised_amount_usd.mean(),2)))
print('Angel - ' + str(round(df_angel.raised_amount_usd.mean(),2)))

In [ ]:
m_venture = df_venture.raised_amount_usd.mean()
m_seed = df_seed.raised_amount_usd.mean()
m_pequity = df_pequity.raised_amount_usd.mean()
m_angel = df_angel.raised_amount_usd.mean()

print(m_venture, m_seed, m_pequity, m_angel)

In [ ]:
# 5 to 15 million
lower = 5000000
upper = 15000000

In [ ]:
m_venture>=lower and m_venture<upper

In [ ]:
m_seed>=lower and m_seed<=upper

In [ ]:
m_pequity>=lower and m_pequity<=upper

In [ ]:
m_angel>=lower and m_angel<=upper

### [Answer] Q. Considering that Spark Funds wants to invest between 5 to 15 million USD per  investment round, which investment type is the most suitable for them?


### Checkpoint 3: Country Analysis

In [ ]:
# From the previous analysis - venture fund is suitable to invest

In [ ]:
df_venture.info()

In [ ]:
df_venture.describe()

In [ ]:
df_venture.head(10)

In [ ]:
df_venture['category_list'] = df_venture['category_list'].astype(str)

In [ ]:
# df_venture =df_venture[pd.notnull(df_venture['country_code'])]

In [ ]:
df_venture.country_code.value_counts().nlargest(9)

In [ ]:
df_venture.columns

### 1. Spark Funds wants to see the top nine countries which have received the highest total funding (across ALL sectors for the chosen investment type)

In [ ]:
df_venture.groupby('country_code')['raised_amount_usd'].sum().sort_values(ascending=False).head(15)

#### Ignoring the unknown country as we dont know what are those

In [ ]:
top9_countries = ['USA','CHN', 'GBR', 'IND', 'CAN','FRA','ISR', 'DEU', 'JPN']

### 2. For the chosen investment type, make a data frame named top9 with the top nine countries (based on the total investment amount each country has received)

In [ ]:
top9 = df_venture[df_venture['country_code'].isin(top9_countries)]

In [ ]:
top9.info()

In [ ]:
top9.head(15)

In [ ]:
top9.describe()

### Identify the top three English-speaking countries in the data frame top9.

In [ ]:
# https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes
list_eng_countries = ['GBR', 'AUS', 'BHS', 'BWA', 'CAN','CYP', 'FJI', 'IND', 
                      'KEN', 'KIR', 'MLT', 'NGA', 'PAK', 'PNG', 'IRL', 'ZAF', 
                      'NZL', 'SGP', 'PHL', 'GMB']
print(len(list_eng_countries))

In [ ]:
countries = 'ATG,AUS,BHS,BRB,BLZ,BWA,CMR,CAN,DMA,ERI,ETH,FJI,GMB,GHA,GRD,GUY,IND,IRL,JAM,KEN,KIR,NLD,LSO,LBR,MWI,MLT,MHL,MUS,FSM,NAM,NRU,NZL,NGA,PAK,PLW,PNG,PHL,RWA,KNA,LCA,VCT,WSM,SYC,SLE,SGP,SLB,ZAF,SSD,SDN,SWZ,TZA,TON,TTO,TUV,GBR,USA,UGA,VUT,ZMB,ZWE'
eng_countries = countries.split(',')
print(len(eng_countries))

In [ ]:
list_eng_countries = list(set(list_eng_countries)|set(eng_countries))
print(type(list_eng_countries), len(list_eng_countries), list_eng_countries)

In [ ]:
list_eng_countries.sort()

In [ ]:
print(list_eng_countries, len(list_eng_countries))

In [ ]:
top9_eng = top9[top9['country_code'].isin(list_eng_countries)]

In [ ]:
top9_eng.country_code.value_counts()

In [ ]:
top9_eng.groupby('country_code')['raised_amount_usd'].sum().sort_values(ascending=False)

### Top3 english speaking countries - Investment wise

#### 1.USA
#### 2. GBR
#### 3. IND

In [ ]:
top9_eng.info()

In [ ]:
top3_country_list = ['USA','GBR','IND']

In [ ]:
top3 = top9_eng[top9_eng['country_code'].isin(top3_country_list)]

In [ ]:
top3.info()

### Checkpoint 4: Sector Analysis 1

In [ ]:
df_mapping = pd.read_csv('mapping.csv',encoding = "ISO-8859-1")

In [ ]:
df_mapping.columns

In [ ]:
df_mapping.info()

In [ ]:
df_mapping.head(10)

In [ ]:
def reCategorize(val):
    splits = val.split('|')
    if(len(splits)>1):
        return splits[0]
    else:
        return val

In [ ]:
print(reCategorize('Apps|Cable|Distribution|Software'))

In [ ]:
print(reCategorize('Analytics'))

In [ ]:
print(reCategorize('Curated Web'))

In [ ]:
# df_venture['primary_sector'] = df_venture['category_list'].apply(reCategorize)

In [ ]:
master_frame['primary_sector'] = master_frame['category_list'].apply(reCategorize)

In [ ]:
master_frame.head()

In [ ]:
master_frame['primary_sector'] = master_frame['primary_sector'].str.lower()

In [ ]:
master_frame.head()

In [ ]:
# primary sector column has been created with first category but not mapped according the mapping.csv; lets do that.

In [ ]:
df_mapping.shape

In [ ]:
df_mapping.head(10)

In [ ]:
df_mapping_melt = pd.melt(df_mapping, id_vars=['category_list'])

In [ ]:
df_mapping_melt.shape

In [ ]:
type(df_mapping_melt)

In [ ]:
df_mapping_melt = df_mapping_melt[df_mapping_melt['value']!=0]

In [ ]:
df_mapping_melt.shape

In [ ]:
df_mapping_melt.head()

In [ ]:
df_mapping_melt.columns

In [ ]:
df_mapping_melt.drop('value', axis=1, inplace=True)

In [ ]:
df_mapping_melt.columns

In [ ]:
df_mapping_melt['category_list'] = df_mapping_melt['category_list'].str.lower()

In [ ]:
df_mapping_melt.rename(columns={'category_list':'primary_sector'}, inplace=True)

In [ ]:
df_mapping_melt.to_csv('mapping_melt.csv')

There are some spelling mistakes in the mapping file on observation; there are 0 present in case where na should be present. so applying regex and correct them

In [ ]:
df_mapping_melt['primary_sector'] = df_mapping_melt.primary_sector.apply(lambda x: re.sub('[0]', 'na', str(x)))

In [ ]:
df_mapping_melt.to_csv('mapping_melt_corrected.csv')

In [ ]:
# master_frame.drop(columns=['_merge','main_sector'], inplace=True)

If we dont rename, it creating two columns with category_list_x and category_list_y; so rename and merge

In [ ]:
master_frame = pd.merge(master_frame, df_mapping_melt, how='left', on='primary_sector',indicator=True)


In [ ]:
master_frame.head()

In [ ]:
round(100*master_frame.isnull().sum()/len(master_frame.index),2)

In [ ]:
master_frame.variable.unique()

In [ ]:
master_frame.rename(columns={'variable':'main_sector'}, inplace=True)

In [ ]:
master_frame.to_csv('master_1.csv')

## Checkpoint 5: Sector Analysis 2

In [ ]:
master_frame.columns

In [ ]:
f_venture = master_frame['funding_round_type']=='venture'
df_venture = master_frame[f_venture]

In [ ]:
df_top3 = df_venture[df_venture['country_code'].isin(top3_country_list)]

In [ ]:
df_top3.head()

In [ ]:
df_top3.info()

In [ ]:
df_top3.shape

In [ ]:
df_top3.country_code.value_counts()

In [ ]:
df_top3.to_csv('top3.csv')

In [ ]:
val2 = 20000000.00
val2>=lower and val2<upper

If we observe there are some investment which are not falling in between 5 and 15 millioin, lets filter out

In [ ]:
# df_top3.drop(df_top3[(df_top3.raised_amount_usd < 5000000)].index, inplace=True)

In [ ]:
 df_top3 = df_top3.drop(df_top3[(df_top3.raised_amount_usd < 5000000)].index)

In [ ]:
 df_top3 = df_top3.drop(df_top3[(df_top3.raised_amount_usd > 15000000)].index)

In [ ]:
df_top3.info()

In [ ]:
df_top3.dropna(subset=['main_sector'], inplace=True)

In [ ]:
df_top3.info()

In [ ]:
df_top3.drop('_merge', axis=1, inplace=True)

In [ ]:
df_top3.head()

#### D1, D2, D3 Dataframe
D1 - USA

D2 - GBR

D3 - IND

In [ ]:
d1 = df_top3[df_top3['country_code'] == 'USA']
d2 = df_top3[df_top3['country_code'] == 'GBR']
d3 = df_top3[df_top3['country_code'] == 'IND']

In [ ]:
print(d1.shape)
print(d2.shape)
print(d3.shape)

In [ ]:
# The total number (or count) of investments for each main sector in a separate column

In [ ]:
d1.columns

### D1

In [ ]:
d1.raised_amount_usd.sum()

In [ ]:
d1.pivot_table(values='raised_amount_usd', index='main_sector',aggfunc={'sum',np.size})

In [ ]:
f_d1_others = d1['main_sector']=='Others'

In [ ]:
d1_others = d1[f_d1_others]

In [ ]:
d1_others.head()

In [ ]:
d1_others.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

In [ ]:
f_d1_sfaa = d1['main_sector']=='Social, Finance, Analytics, Advertising'
d1_sfaa = d1[f_d1_sfaa]
d1_sfaa.head()

In [ ]:
d1_sfaa.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

### D2

In [ ]:
d2.raised_amount_usd.sum()

In [ ]:
d2.pivot_table(values='raised_amount_usd', index='main_sector',aggfunc={'sum',np.size})

In [ ]:
f_d2_others = d2['main_sector']=='Others'
d2_others = d2[f_d2_others]
d2_others.head()

In [ ]:
d2_others.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

In [ ]:
f_d2_sfaa = d2['main_sector']=='Social, Finance, Analytics, Advertising'
d2_sfaa = d2[f_d2_sfaa]
d2_sfaa.head()
d2_sfaa.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

### D3

In [ ]:
d3.raised_amount_usd.sum()

In [ ]:
d3.pivot_table(values='raised_amount_usd', index='main_sector',aggfunc={'sum',np.size})

In [ ]:
f_d3_cs = d3['main_sector']=='Others'
d3_cs = d3[f_d3_cs]
d3_cs.head()
d3_cs.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

In [ ]:
f_d3_others = d3['main_sector']=='Social, Finance, Analytics, Advertising'
d3_others = d3[f_d3_others]
d3_others.head()
d3_others.groupby('permalink')['raised_amount_usd'].sum().sort_values(ascending=False).head()

## Checkpoint 6: Plots

In [ ]:
import matplotlib.pyplot as plt

%matplotlib inline

import seaborn as sns


In [ ]:
list_interested_fundtypes = ['venture', 'angel', 'seed', 'private_equity']

In [ ]:
df_fundtypes_vasp = master_frame[master_frame['funding_round_type'].isin(list_interested_fundtypes)]

In [ ]:
df_fundtypes_vasp.groupby('funding_round_type').sum()

In [ ]:
df_fundtypes_vasp.groupby('funding_round_type').mean()

In [ ]:
df_fundtypes_vasp.groupby('funding_round_type').median()

In [ ]:
plt.figure(figsize=(10,10))
sns.barplot(x='funding_round_type', y='raised_amount_usd', data=df_fundtypes_vasp)
plt.axhline(5000000, linestyle='--',label='5 million')
plt.axhline(15000000, linestyle='--', label='15 million')
# ax1.axhline(10, ls='--')
# ax2.axhline(30, ls='--')

# ax1.text(0.5,25, "Some text")
# ax2.text(0.5,25, "Some text")
plt.show()

In [ ]:
plt.figure(figsize=(30, 20))
plt.subplot(2, 1, 1)
sns.countplot(x="country_code", data=top9)
plt.title("Count")
plt.subplot(2, 1, 2)
sns.barplot(x="country_code", y="raised_amount_usd", data=top9, estimator=sum)
plt.title("Sum")
plt.show()

In [ ]:
# plt.figure(figsize=(25, 15))
# plt.subplot(2, 1, 1)
# sns.countplot(x="country_code", data=top9)
# plt.title("Count")
# plt.subplot(2, 1, 2)
# sns.barplot(x="country_code", y="raised_amount_usd", data=top9, estimator=np.mean)
# plt.title("Sum")
# plt.show()

In [ ]:
plt.figure(figsize=(30, 20))
sns.barplot(x='raised_amount_usd', y='country_code', hue="main_sector", data=df_top3, estimator=np.sum)
plt.show()

In [ ]:
d1_top3_sectors_list = ["Others","Social, Finance, Analytics, Advertising", "Cleantech / Semiconductors"]
d1_top3 = d1[d1['main_sector'].isin(d1_top3_sectors_list)]
# d1_top3.head()

In [ ]:
d2_top3_sectors_list = ["Others","Social, Finance, Analytics, Advertising", "Cleantech / Semiconductors"]
d2_top3 = d2[d2['main_sector'].isin(d2_top3_sectors_list)]
# d2_top3.head()

In [ ]:
d3_top3_sectors_list = ["Others","Social, Finance, Analytics, Advertising", "News, Search and Messaging"]
d3_top3 = d3[d3['main_sector'].isin(d3_top3_sectors_list)]
d3_top3.head()

In [ ]:
df_top3_sectors = pd.concat([d1_top3, d2_top3, d3_top3])

In [ ]:
df_top3_sectors.country_code.unique()

In [ ]:
df_top3_sectors.shape

In [ ]:
plt.figure(figsize=(30, 20))
# sns.barplot(x='raised_amount_usd', y='country_code', hue="main_sector", data=df_top3_sectors, estimator=np.sum, cmap=pal)
sns.barplot(x='raised_amount_usd', y='country_code', hue="main_sector", data=df_top3_sectors, estimator=np.sum)
plt.show()